// astro/dist/actions/consts.js
const ACTIONS_TYPES_FILE = "actions.d.ts";
const VIRTUAL_MODULE_ID = "astro:actions";
const RESOLVED_VIRTUAL_MODULE_ID = "\0" + VIRTUAL_MODULE_ID;
const RUNTIME_VIRTUAL_MODULE_ID = "virtual:astro:actions/runtime";
const RESOLVED_RUNTIME_VIRTUAL_MODULE_ID = "\0" + RUNTIME_VIRTUAL_MODULE_ID;
const ENTRYPOINT_VIRTUAL_MODULE_ID = "virtual:astro:actions/entrypoint";
const RESOLVED_ENTRYPOINT_VIRTUAL_MODULE_ID = "\0" + ENTRYPOINT_VIRTUAL_MODULE_ID;
const OPTIONS_VIRTUAL_MODULE_ID = "virtual:astro:actions/options";
const RESOLVED_OPTIONS_VIRTUAL_MODULE_ID = "\0" + OPTIONS_VIRTUAL_MODULE_ID;
const RESOLVED_NOOP_ENTRYPOINT_VIRTUAL_MODULE_ID = "\0virtual:astro:actions/noop-entrypoint";
const ACTION_QUERY_PARAMS = {
  actionName: "_action",
  actionPayload: "_astroActionPayload"
};
const ACTION_RPC_ROUTE_PATTERN = "/_actions/[...path]";
export {
  ACTIONS_TYPES_FILE,
  ACTION_QUERY_PARAMS,
  ACTION_RPC_ROUTE_PATTERN,
  ENTRYPOINT_VIRTUAL_MODULE_ID,
  OPTIONS_VIRTUAL_MODULE_ID,
  RESOLVED_ENTRYPOINT_VIRTUAL_MODULE_ID,
  RESOLVED_NOOP_ENTRYPOINT_VIRTUAL_MODULE_ID,
  RESOLVED_OPTIONS_VIRTUAL_MODULE_ID,
  RESOLVED_RUNTIME_VIRTUAL_MODULE_ID,
  RESOLVED_VIRTUAL_MODULE_ID,
  RUNTIME_VIRTUAL_MODULE_ID,
  VIRTUAL_MODULE_ID
};
